### PR TITLE
feature: tree dnd

### PR DESCRIPTION
**Pull request by Kupov**
https://jira.railsc.ru/browse/GOODS-488
https://jira.railsc.ru/browse/GOODS-503

### CODE REVIEW

File: src/SaveControl/sagas.js [hunk 1]
```diff
@@ -0,0 +1,257 @@
...
+    yield put({
+      type: SAVE_DIFF,
+      payload: {
+        waitingState,
+        prevState: action.payload.curState
+      }
+    });
+  } catch (err) {
+    console.log(err);
+  }
```
Comment (apress-hound) [inline src/SaveControl/sagas.js]: Unrecoverable syntax error. (81% scanned).

File: src/SaveControl/sagas.js [hunk 2]
```diff
@@ -0,0 +1,257 @@
...
+    }
+
+    yield put({
+      type: SAVE_DIFF,
+      payload: {
+        waitingState,
+        prevState: action.payload.curState
+      }
+    });
+  } catch (err) {
```
Comment (apress-hound) [inline src/SaveControl/sagas.js]: Expected an identifier and instead saw 'catch'.<br>Expected an operator and instead saw '('.<br>Missing semicolon.<br>Expected an identifier and instead saw ')'.

File: src/SaveControl/sagas.js [hunk 3]
```diff
@@ -0,0 +1,257 @@
...
+              waitingItem[key] = {...waitingItem[key], ...record[key]};
+            } else {
+              waitingItem[key] = record[key];
+            }
+          });
+        } else {
+          waitingState.push(record);
+        }
+      });
+    } else {
```
Comment (apress-hound) [inline src/SaveControl/sagas.js]: Expected 'catch' and instead saw 'else'.<br>Expected an identifier and instead saw 'else'.<br>Missing semicolon.

File: src/SaveControl/sagas.js [hunk 4]
```diff
@@ -0,0 +1,257 @@
...
+            if (typeof record[key] === 'object' && record[key] !== null) {
+              waitingItem[key] = {...waitingItem[key], ...record[key]};
+            } else {
+              waitingItem[key] = record[key];
+            }
+          });
+        } else {
+          waitingState.push(record);
+        }
+      });
```
Comment (apress-hound) [inline src/SaveControl/sagas.js]: Expected an identifier and instead saw ')'.

File: src/SaveControl/sagas.js [hunk 5]
```diff
@@ -0,0 +1,257 @@
...
+          }
+
+          Object.keys(record).forEach((key) => {
+            if (typeof record[key] === 'object' && record[key] !== null) {
+              waitingItem[key] = {...waitingItem[key], ...record[key]};
+            } else {
+              waitingItem[key] = record[key];
+            }
+          });
+        } else {
```
Comment (apress-hound) [inline src/SaveControl/sagas.js]: Expected ')' and instead saw 'else'.<br>Missing semicolon.

File: src/SaveControl/sagas.js [hunk 6]
```diff
@@ -0,0 +1,257 @@
...
+            delete waitingItem.destroy;
+          }
+
+          Object.keys(record).forEach((key) => {
+            if (typeof record[key] === 'object' && record[key] !== null) {
+              waitingItem[key] = {...waitingItem[key], ...record[key]};
+            } else {
+              waitingItem[key] = record[key];
+            }
+          });
```
Comment (apress-hound) [inline src/SaveControl/sagas.js]: Expected an identifier and instead saw ')'.

File: src/SaveControl/sagas.js [hunk 7]
```diff
@@ -0,0 +1,257 @@
...
+
+        if (waitingItem) {
+          if (waitingItem.destroy && !record.destroy) {
+            delete waitingItem.destroy;
+          }
+
+          Object.keys(record).forEach((key) => {
+            if (typeof record[key] === 'object' && record[key] !== null) {
+              waitingItem[key] = {...waitingItem[key], ...record[key]};
+            } else {
```
Comment (apress-hound) [inline src/SaveControl/sagas.js]: Expected ')' and instead saw 'else'.<br>Missing semicolon.

File: src/SaveControl/sagas.js [hunk 8]
```diff
@@ -0,0 +1,257 @@
...
+        const waitingItem = waitingState.find(item => item.id === record.id);
+
+        if (waitingItem) {
+          if (waitingItem.destroy && !record.destroy) {
+            delete waitingItem.destroy;
+          }
+
+          Object.keys(record).forEach((key) => {
+            if (typeof record[key] === 'object' && record[key] !== null) {
+              waitingItem[key] = {...waitingItem[key], ...record[key]};
```
Comment (apress-hound) [inline src/SaveControl/sagas.js]: Expected '}' to match '{' from line 187 and instead saw '...'.<br>Missing semicolon.<br>Unnecessary semicolon.

File: src/Filter/import.js [hunk 9]
```diff
@@ -0,0 +1,19 @@
...
+  Component,
+  connect,
+  browserHistory,
+
+  actionsTree,
+  actionsTable,
+  actionsFilter,
+
+  put,
+};
```
Comment (apress-hound) [inline src/Filter/import.js]: Expected an identifier and instead saw '}'.<br>Unexpected ';'.<br>'}' is not defined.

File: dist/assets-help.bundle.js [hunk 10]
```diff
@@ -10,540 +10,540 @@
...
+/******/ 		var me = installedModules[moduleId];
+/******/ 		if(!me) return __webpack_require__;
+/******/ 		var fn = function(request) {
+/******/ 			if(me.hot.active) {
+/******/ 				if(installedModules[request]) {
+/******/ 					if(installedModules[request].parents.indexOf(moduleId) < 0)
+/******/ 						installedModules[request].parents.push(moduleId);
+/******/ 					if(me.children.indexOf(request) < 0)
+/******/ 						me.children.push(request);
+/******/ 				} else hotCurrentParents = [moduleId];
```
Comment (apress-hound) [inline dist/assets-help.bundle.js]: Expected '{' and instead saw 'hotCurrentParents'.<br>Too many errors. (2% scanned).

File: src/Filter/Filter.jsx [hunk 11]
```diff
@@ -1,27 +1,37 @@
-import React, {Component} from 'react';
-import {browserHistory} from 'react-router';
+import {
```
Comment (dontuse) [inline src/Filter/Filter.jsx]: ну и в чем от этого бенефит ? я бы предложил делать файл export.js в котором бы лежали все необходимые импортируемые сущности, filter.Filter, filter.sagas, filter.actions .... вообщем я не понял что это дает.

File: src/App/App.jsx [hunk 12]
```diff
@@ -28,6 +28,22 @@ import {block} from '../utils';
 import '../styles/global.scss';
 import './style.scss';
 
+window.app = {
```
Comment (Kupov) [inline src/App/App.jsx]: ок

File: src/Filter/Filter.jsx (hunk 11 above)
Comment (Kupov) [inline src/Filter/Filter.jsx]: уже обсудили

File: src/Tree/ContainerTree.jsx [hunk 13]
```diff
@@ -0,0 +1,68 @@
...
+class ContainerTree extends Component {
+  static propTypes = {
+    tree: PropTypes.array,
+  };
+
+  shouldComponentUpdate(nextProps) {
+    return isChangeObj(this.props, nextProps);
+  }
+
+  actionMoveNode = (...args) => this.props.dispatch(actions.moveNode(...args))
```
Comment (dontuse) [inline src/Tree/ContainerTree.jsx]: переносы строк должны быть между методами

File: src/utils/index.js [hunk 14]
```diff
@@ -31,3 +31,33 @@ export function checkStatus(response) {
   error.response = response;
   throw error;
 }
+
+export function isChangeObj(objA, objB) {
```
Comment (dontuse) [inline src/utils/index.js]: _.isEqual(value, other)  у нас есть лодаш, чем не подходит эта ф-ия https://lodash.com/docs/#isEqual ?

File: src/Tree/reducer.js [hunk 15]
```diff
@@ -36,6 +42,110 @@ const isSelected = (treeState, action) => {
...
+        };
+
+        if (childNode.moveNode) {
+          moveNode = childNode.moveNode;
+
+          newNode = {
+            ...newNode,
+            expandable: !!newNode['tree_nodes'].length,
+            items_count: newNode['items_count'] - moveNode['items_count'],
+          };
```
Comment (apress-hound) [inline src/Tree/reducer.js]: Unexpected '}'.

File: src/Tree/reducer.js [hunk 16]
```diff
@@ -36,6 +42,110 @@ const isSelected = (treeState, action) => {
   return selected.reverse();
 };
 
+
+const deleteNode = (state, id) => {
+  let newState = [];
```
Comment (apress-hound) [inline src/Tree/reducer.js]: 'newState' is defined but never used.

File: src/Tree/reducer.js [hunk 17]
```diff
@@ -36,6 +42,110 @@ const isSelected = (treeState, action) => {
...
+          tree_nodes: childNode.newState
+        };
+
+        if (childNode.moveNode) {
+          moveNode = childNode.moveNode;
+
+          newNode = {
+            ...newNode,
+            expandable: !!newNode['tree_nodes'].length,
+            items_count: newNode['items_count'] - moveNode['items_count'],
```
Comment (apress-hound) [inline src/Tree/reducer.js]: Identifier 'items_count' is not in camel case.<br>Missing semicolon.<br>Expected '}' to match '{' from line 46 and instead saw ':'.<br>Unrecoverable syntax error. (27% scanned).

File: src/Tree/reducer.js [hunk 18]
```diff
@@ -36,6 +42,110 @@ const isSelected = (treeState, action) => {
...
+          ...node,
+          tree_nodes: childNode.newState
+        };
+
+        if (childNode.moveNode) {
+          moveNode = childNode.moveNode;
+
+          newNode = {
+            ...newNode,
+            expandable: !!newNode['tree_nodes'].length,
```
Comment (apress-hound) [inline src/Tree/reducer.js]: Missing semicolon.<br>Expected '}' to match '{' from line 63 and instead saw ':'.

File: src/Tree/reducer.js [hunk 19]
```diff
@@ -36,6 +42,110 @@ const isSelected = (treeState, action) => {
...
+        newNode = {
+          ...node,
+          tree_nodes: childNode.newState
+        };
+
+        if (childNode.moveNode) {
+          moveNode = childNode.moveNode;
+
+          newNode = {
+            ...newNode,
```
Comment (apress-hound) [inline src/Tree/reducer.js]: Expected '}' to match '{' from line 66 and instead saw '...'.<br>Missing semicolon.

File: src/Tree/reducer.js [hunk 20]
```diff
@@ -36,6 +42,110 @@ const isSelected = (treeState, action) => {
...
+      moveNode = node;
+    } else {
+      let newNode = {...node};
+
+      if (node.tree_nodes && node.tree_nodes.length) {
+        const childNode = deleteNode(node.tree_nodes, id);
+        newNode = {
+          ...node,
+          tree_nodes: childNode.newState
+        };
```
Comment (apress-hound) [inline src/Tree/reducer.js]: Expected ')' and instead saw ';'.<br>Missing semicolon.

File: src/Tree/reducer.js [hunk 21]
```diff
@@ -36,6 +42,110 @@ const isSelected = (treeState, action) => {
...
+    if (node.id === id) {
+      moveNode = node;
+    } else {
+      let newNode = {...node};
+
+      if (node.tree_nodes && node.tree_nodes.length) {
+        const childNode = deleteNode(node.tree_nodes, id);
+        newNode = {
+          ...node,
+          tree_nodes: childNode.newState
```
Comment (apress-hound) [inline src/Tree/reducer.js]: Identifier 'tree_nodes' is not in camel case.<br>Missing semicolon.<br>Expected '}' to match '{' from line 56 and instead saw ':'.

File: src/Tree/reducer.js [hunk 22]
```diff
@@ -36,6 +42,110 @@ const isSelected = (treeState, action) => {
...
+  state.forEach((node) => {
+    if (node.id === id) {
+      moveNode = node;
+    } else {
+      let newNode = {...node};
+
+      if (node.tree_nodes && node.tree_nodes.length) {
+        const childNode = deleteNode(node.tree_nodes, id);
+        newNode = {
+          ...node,
```
Comment (apress-hound) [inline src/Tree/reducer.js]: Expected '}' to match '{' from line 58 and instead saw '...'.<br>Missing semicolon.

File: src/Tree/reducer.js [hunk 23]
```diff
@@ -36,6 +42,110 @@ const isSelected = (treeState, action) => {
...
+  let moveNode = null;
+
+  state.forEach((node) => {
+    if (node.id === id) {
+      moveNode = node;
+    } else {
+      let newNode = {...node};
+
+      if (node.tree_nodes && node.tree_nodes.length) {
+        const childNode = deleteNode(node.tree_nodes, id);
```
Comment (apress-hound) [inline src/Tree/reducer.js]: Identifier 'tree_nodes' is not in camel case.<br>'childNode' is defined but never used.

File: src/Tree/reducer.js [hunk 24]
```diff
@@ -36,6 +42,110 @@ const isSelected = (treeState, action) => {
...
+  let newState = [];
+  let moveNode = null;
+
+  state.forEach((node) => {
+    if (node.id === id) {
+      moveNode = node;
+    } else {
+      let newNode = {...node};
+
+      if (node.tree_nodes && node.tree_nodes.length) {
```
Comment (apress-hound) [inline src/Tree/reducer.js]: Identifier 'tree_nodes' is not in camel case.

**Comment by dontuse** [comment]
не делай так больше, все правки в таких огромных коммитах без амендов

**Pull request merged**